File: frontend/src/api/shader.ts
import { apiFetch, parseApiError, resolveApiUrl } from "./client";

export type QualityPreset = "fast" | "balanced" | "high" | "manual";

export interface MinPipelineTracePhase {
  phase: string;
  status: string;
  duration_ms?: number | null;
  message?: string | null;
}

export interface ShaderGraphShadowSummary {
  status: "rendered" | "unsupported" | "failed";
  renderer_path?: string | null;
  dsl_schema_version?: string | null;
  compiler_version?: string | null;
  document_sha256?: string | null;
  topology_sha256?: string | null;
  layer_count?: number | null;
  primitive_count?: number | null;
  compile_count?: number | null;
  cache_hit_count?: number | null;
  cache_size?: number | null;
  render_duration_ms?: number | null;
  unsupported_features?: string[] | null;
  error_code?: string | null;
  shader_graph?: {
    schema_version?: string;
    layers?: Array<Record<string, unknown>>;
    [key: string]: unknown;
  } | null;
}

// scene_mvp 最小管线的运行摘要；后端字段可缺省，前端全部按可选处理。
export interface MinPipelineSummary {
  mae?: number | null;
  objective_loss?: number | null;
  metric_breakdown?: {
    metric_version?: string;
    global_mae?: number;
    foreground_mae?: number;
    background_mae?: number;
    geometry_mask_loss?: number;
    edge_loss?: number;
    worst_tile_mae?: number;
  } | null;
  template_version?: string | null;
  render_count?: number | null;
  render_budget?: number | null;
  llm_call_count?: number | null;
  llm_budget?: number | null;
  refine_budget?: number | null;
  scene?: unknown;
  trace?: MinPipelineTracePhase[] | null;
  // 质量目标：区分“流程完成”与“质量达标”
  target_mae?: number | null;
  target_loss?: number | null;
  target_reached?: boolean | null;
  // prepared 渲染路径与 prepare 阶段耗时
  renderer_path?: string | null;
  prepare_duration_ms?: number | null;
  // prepared 后复用 uniform 的热渲染统计
  uniform_render_count?: number | null;
  uniform_render_p95_ms?: number | null;
  shader_graph_shadow?: ShaderGraphShadowSummary | null;
}

export interface ShaderResponse {
  project_id: string;
  run_id: string;
  glsl: string;
  generation_mode: "scene_mvp";
  quality_preset: QualityPreset;
  stop_reason?: string | null;
  render_width?: number | null;
  render_height?: number | null;
  final_render_url?: string | null;
  metrics_url?: string | null;
  manifest_url?: string | null;
  min_pipeline: MinPipelineSummary;
}

export interface ShaderApiFailure {
  status: number;
  code?: string;
  runId?: string;
  stage?: string;
  retryable?: boolean;
  stopReason?: string;
}

export class ShaderApiError extends Error {
  readonly failure: ShaderApiFailure;

  constructor(message: string, failure: ShaderApiFailure) {
    super(message);
    this.name = "ShaderApiError";
    this.failure = failure;
  }
}

export function resolveShaderApiUrl(path: string): string {
  return resolveApiUrl(path);
}

function readString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value : undefined;
}

async function readError(response: Response, fallback: string): Promise<ShaderApiError> {
  const parsed = await parseApiError(response, fallback);
  return new ShaderApiError(parsed.message, {
    status: parsed.status,
    code: readString(parsed.fields.code),
    runId: readString(parsed.fields.run_id),
    stage: readString(parsed.fields.stage),
    retryable:
      typeof parsed.fields.retryable === "boolean"
        ? parsed.fields.retryable
        : undefined,
    stopReason: readString(parsed.fields.stop_reason),
  });
}

export interface GenerateShaderOptions {
  runId?: string;
  qualityPreset: QualityPreset;
  instruction: string;
  signal?: AbortSignal;
}

export async function generateShader(
  file: File,
  options: GenerateShaderOptions,
): Promise<ShaderResponse> {
  const formData = new FormData();
  formData.append("file", file);
  if (options.runId) formData.append("run_id", options.runId);
  formData.append("quality_preset", options.qualityPreset);
  formData.append("instruction", options.instruction);

  const response = await apiFetch("/api/shader/generate", {
    method: "POST",
    body: formData,
    signal: options.signal,
  });

  if (!response.ok) {
    throw await readError(response, "生成 GLSL 失败。");
  }

  return response.json();
}

// scene_mvp 运行中单节点进度事件；后端白名单字段，全部按可缺省处理。
export interface MinRunProgressEvent {
  seq: number;
  node: string;
  status: string;
  phase?: string | null;
  elapsed_ms?: number | null;
  duration_ms?: number | null;
  budgets?: Record<string, number> | null;
  counters?: {
    render_count?: number;
    llm_call_count?: number;
    refine_count?: number;
  } | null;
  best?: { mae?: number; loss?: number } | null;
  trace?: Array<Record<string, unknown>> | null;
  next_action?: string | null;
  stop_reason?: string | null;
}

export interface MinRunProgressSnapshot {
  budgets?: Record<string, number> | null;
  counters?: {
    render_count?: number;
    llm_call_count?: number;
    refine_count?: number;
  } | null;
  best?: { mae?: number; loss?: number } | null;
  current_node?: string | null;
  render_seq?: number | null;
}

export type MinRunProgressStatus = "pending" | "running" | "succeeded" | "failed";

export interface MinRunProgressResponse {
  run_id: string;
  status: MinRunProgressStatus;
  generation_mode?: string | null;
  quality_preset?: string | null;
  started_at?: string | null;
  latest_seq: number;
  events: MinRunProgressEvent[];
  snapshot: MinRunProgressSnapshot;
}

export async function fetchMinRunProgress(
  runId: string,
  after = 0,
): Promise<MinRunProgressResponse> {
  const query = after > 0 ? `?after=${Math.floor(after)}` : "";
  const response = await apiFetch(
    `/api/shader/runs/${encodeURIComponent(runId)}/progress${query}`,
  );
  if (!response.ok) {
    throw await readError(response, "读取运行进度失败。");
  }
  return response.json();
}

export function resolveMinRunRenderUrl(runId: string, renderSeq: number): string {
  return resolveApiUrl(
    `/api/shader/runs/${encodeURIComponent(runId)}/progress/render?seq=${renderSeq}`,
  );
}

File: backend/app/schemas/shader.py
"""Shader API schema."""

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

GenerationMode = Literal["scene_mvp"]
QualityPresetName = Literal["fast", "balanced", "high", "manual"]
ShaderEngineId = Literal["shader_graph_v1", "direct_glsl_layerplan_v1"]
ShaderRepresentation = Literal["shader_document_v1", "shader_program_spec_v1"]


class ShaderEngineAttemptSummary(BaseModel):
    """父 run 可公开的单个 child attempt 安全引用."""

    attempt_id: str
    engine: ShaderEngineId
    representation: ShaderRepresentation
    status: Literal["succeeded", "failed"]
    failure_code: str | None = None


class ShaderShadowSubmissionSummary(BaseModel):
    """production shadow 的非权威提交结果；不代表 attempt 执行成功."""

    model_config = ConfigDict(extra="allow")

    status: str
    reason: str
    attempt_id: str | None = None


class ShaderEngineRunSummary(BaseModel):
    """父 run 的 policy、attempt 与 fallback 安全摘要."""

    model_config = ConfigDict(extra="allow")

    policy_id: str
    policy_sha256: str
    configured_stage: str
    stage: str
    bucket: int | None = None
    selected_attempt_id: str
    attempt_refs: list[ShaderEngineAttemptSummary] = Field(default_factory=list)
    fallback_from: ShaderEngineId | None = None
    fallback_reason: str | None = None
    promotion_authorization_sha256: str | None = None
    shadow_submission: ShaderShadowSubmissionSummary | None = None


class ShaderGraphShadowSummary(BaseModel):
    """不参与产品选择的 ShaderGraph shadow 纵向切片摘要."""

    status: Literal["rendered", "unsupported", "failed"]
    renderer_path: Literal["compiled_graph_program_cache_v1"] | None = None
    dsl_schema_version: str | None = None
    compiler_version: str | None = None
    document_sha256: str | None = None
    topology_sha256: str | None = None
    layer_count: int = 0
    primitive_count: int = 0
    compile_count: int = 0
    cache_hit_count: int = 0
    cache_size: int = 0
    render_duration_ms: float | None = None
    unsupported_features: list[str] = Field(default_factory=list)
    error_code: str | None = None
    resource_summary: dict[str, int] = Field(default_factory=dict)
    shader_graph: dict[str, Any] | None = None


class ShaderMinPipelineSummary(BaseModel):
    """scene_mvp 最小流水线的公开运行摘要."""

    mae: float | None = None
    objective_loss: float | None = None
    metric_breakdown: dict[str, Any] = Field(default_factory=dict)
    template_version: str
    render_count: int = 0
    render_budget: int
    llm_call_count: int = 0
    llm_budget: int
    refine_budget: int
    run_classification: Literal["frozen_benchmark", "independent_experiment"]
    experiment_id: str | None = None
    config_fingerprint: str
    report_schema_version: str
    patch_candidate_draw_budget: int
    patch_evidence: list[dict[str, Any]] = Field(default_factory=list)
    renderer_path: Literal[
        "prepared_uniforms_v1",
        "compiled_graph_program_cache_v1",
        "direct_program_spec_v1",
    ]
    target_mae: float
    target_loss: float
    target_reached: bool
    prepare_duration_ms: float
    uniform_render_count: int
    uniform_render_p95_ms: float
    scene: dict[str, Any] | None = None
    trace: list[dict[str, Any]] = Field(default_factory=list)
    shader_graph_shadow: ShaderGraphShadowSummary | None = None


class ShaderResponse(BaseModel):
    """scene_mvp 产品生成响应."""

    project_id: UUID
    run_id: UUID
    glsl: str
    generation_mode: GenerationMode
    quality_preset: QualityPresetName
    engine: ShaderEngineId | None = None
    representation: ShaderRepresentation | None = None
    engine_run: ShaderEngineRunSummary | None = None
    stop_reason: str | None = None
    render_width: int | None = None
    render_height: int | None = None
    final_render_url: str | None = None
    metrics_url: str | None = None
    manifest_url: str | None = None
    min_pipeline: ShaderMinPipelineSummary


class ShaderGenerationErrorDetail(BaseModel):
    """Shader 生成失败的稳定、可安全公开诊断."""

    message: str
    code: str
    run_id: UUID
    stage: str
    retryable: bool
    stop_reason: str | None = None


class ShaderGenerationErrorResponse(BaseModel):
    """兼容 FastAPI detail envelope 的生成失败响应."""

    detail: ShaderGenerationErrorDetail


class MinRunProgressEvent(BaseModel):
    """scene_mvp 单节点进度事件（白名单，不含图片、Scene 或 GLSL）."""

    model_config = ConfigDict(extra="allow")

    seq: int
    node: str
    status: str
    phase: str | None = None
    elapsed_ms: float | None = None
    duration_ms: float | None = None
    budgets: dict[str, Any] = Field(default_factory=dict)
    counters: dict[str, int] = Field(default_factory=dict)
    best: dict[str, float] = Field(default_factory=dict)
    trace: list[dict[str, Any]] = Field(default_factory=list)
    next_action: str | None = None
    stop_reason: str | None = None


class MinRunProgressSnapshot(BaseModel):
    """运行进度快照：最新计数、质量、当前节点与渲染帧序号."""

    model_config = ConfigDict(extra="allow")

    budgets: dict[str, Any] = Field(default_factory=dict)
    counters: dict[str, int] = Field(default_factory=dict)
    best: dict[str, float] = Field(default_factory=dict)
    current_node: str | None = None
    render_seq: int = 0


class MinRunProgressResponse(BaseModel):
    """scene_mvp 运行进度增量读取响应；未知 run_id 返回 pending."""

    run_id: UUID
    status: Literal["pending", "running", "succeeded", "failed"]
    generation_mode: str | None = None
    quality_preset: str | None = None
    started_at: str | None = None
    latest_seq: int = 0
    events: list[MinRunProgressEvent] = Field(default_factory=list)
    snapshot: MinRunProgressSnapshot = Field(default_factory=MinRunProgressSnapshot)
